subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultConvention.java

```diff
@@ -91,44 +105,68 @@ private Instantiator getInstantiator() {
         return values.get(0);
     }
 
+    @Override
     public void add(String name, Object extension) {
         if (extension instanceof Class) {
             create(name, (Class<?>) extension);
         } else {
-            extensionsStorage.add(name, extension);
+            add(name, (Class) extension.getClass(), extension);
         }
     }
 
+    @Override
+    public <P, I extends P> void add(String name, Class<P> publicType, I extension) {
+        extensionsStorage.add(name, publicType, extension);
+    }
+
+    @Override
     public <T> T create(String name, Class<T> type, Object... constructionArguments) {
-        T instance = getInstantiator().newInstance(type, constructionArguments);
-        add(name, instance);
+        return create(name, type, type, constructionArguments);
+    }
+
+    @Override
+    public <P, I extends P> I create(String name, Class<P> publicType, Class<I> instanceType, Object... constructionArguments) {
```

Comment:
This signature gives different semantics to existing invocations such as `create("name", MyExtensionType.class, ClassInConstructionArgumentPosition.class)`. Moving the `publicType` to the front of the signature would prevent that, what do you think?